src/engine/ult.c

```diff
@@ -610,3 +610,227 @@ dss_main_exec(void (*func)(void *), void *arg)
+	struct dss_module_info *info = dss_get_module_info();
+	int                     xs_id;
+	struct dss_xstream     *dx;
+	struct dss_chore_queue *queue;
+
+	chore->cho_status = DSS_CHORE_NEW;
+	chore->cho_func   = func;
+
+	/*
+	 * The dss_chore_queue_ult approach may get insufficient scheduling on
+	 * a "main" xstream when the chore queue is long. So we fall back to
+	 * the one-ULT-per-chore approach if there's no helper xstream.
+	 */
+	if (dss_tgt_offload_xs_nr == 0) {
+		D_INIT_LIST_HEAD(&chore->cho_link);
+		return dss_ult_create(dss_chore_ult, chore, DSS_XS_IOFW, info->dmi_tgt_id,
+				      0 /* stack_size */, NULL /* ult */);
+	}
+
+	/* Find the chore queue. */
+	xs_id = sched_ult2xs(DSS_XS_IOFW, info->dmi_tgt_id);
+	D_ASSERT(xs_id != -DER_INVAL);
+	dx = dss_get_xstream(xs_id);
+	D_ASSERT(dx != NULL);
+	queue = dx->dx_chore_queue;
+	D_ASSERT(queue != NULL);
+
+	ABT_mutex_lock(queue->chq_mutex);
+	if (queue->chq_stop) {
+		ABT_mutex_unlock(queue->chq_mutex);
+		return -DER_CANCELED;
+	}
+	d_list_add_tail(&chore->cho_link, &queue->chq_list);
+	ABT_cond_broadcast(queue->chq_cond);
+	ABT_mutex_unlock(queue->chq_mutex);
+
+	D_DEBUG(DB_TRACE, "%p: tgt_id=%d -> xs_id=%d dx.tgt_id=%d\n", chore, info->dmi_tgt_id,
+		xs_id, dx->dx_tgt_id);
+	return 0;
+}
+
+/**
+ * Do \a chore for \a func synchronously in the current ULT.
+ *
+ * \param[in]	chore	embedded chore object
+ * \param[in]	func	function to be executed via \a chore
+ */
+void
+dss_chore_diy(struct dss_chore *chore, dss_chore_func_t func)
+{
+	D_INIT_LIST_HEAD(&chore->cho_link);
+	chore->cho_status = DSS_CHORE_NEW;
+	chore->cho_func   = func;
+
+	dss_chore_diy_internal(chore);
+}
+
+static void
+dss_chore_queue_ult(void *arg)
+{
+	struct dss_chore_queue *queue = arg;
+	d_list_t                list  = D_LIST_HEAD_INIT(list);
+
+	D_ASSERT(queue != NULL);
+	D_DEBUG(DB_TRACE, "begin\n");
+
+	for (;;) {
+		struct dss_chore *chore;
+		struct dss_chore *chore_tmp;
+		bool              stop = false;
+
+		/*
+		 * The scheduling order shall be
+		 *
+		 *   [queue->chq_list] [list],
+		 *
+		 * where list contains chores that have returned
+		 * DSS_CHORE_READY in the previous iteration.
+		 */
+		ABT_mutex_lock(queue->chq_mutex);
+		for (;;) {
+			if (!d_list_empty(&queue->chq_list)) {
+				d_list_splice_init(&queue->chq_list, &list);
+				break;
+			}
+			if (!d_list_empty(&list))
+				break;
+			if (queue->chq_stop) {
+				stop = true;
+				break;
+			}
+			sched_cond_wait(queue->chq_cond, queue->chq_mutex);
+		}
+		ABT_mutex_unlock(queue->chq_mutex);
+
+		if (stop)
+			break;
+
+		d_list_for_each_entry_safe(chore, chore_tmp, &list, cho_link) {
+			bool is_reentrance = (chore->cho_status == DSS_CHORE_READY);
+
+			D_DEBUG(DB_TRACE, "%p: status=%d\n", chore, chore->cho_status);
+			chore->cho_status = chore->cho_func(chore, is_reentrance);
+			D_ASSERT(chore->cho_status != DSS_CHORE_NEW);
+			if (chore->cho_status == DSS_CHORE_DONE)
+				d_list_del_init(&chore->cho_link);
+			ABT_thread_yield();
```

Comment:
If the cho_func() return DSS_CHORE_READY, then the chore is still in the temporary list, who and when will insert it back to the queue->chq_list?

Reply:
The `splice` on line 736. Or, have I misunderstood the question?

Reply:
The d_list_splice_init() is for getting chores from queue->chq_list to the temporary list. That means once some chore is moved from queue->chq_list to the temporary list, then it will never be added back to queue->chq_list, right?

Reply:
No, `list` is not cleared before the `splice`. Have I overlooked anything?

Reply:
OK, that matches my understand. Means that the new added chores will always have higher priority to be scheduled. I am not sure whether it is reasonable or not, but large task (with multiple steps) will be delayed more under such schedule policy. It is suggested to verify the behavior under heavy loaded system on large scaled cluster.